edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs

```diff
@@ -32,102 +33,167 @@ public async Task GetDevicesAndModulesInTargetDeviceScope([FromRoute] string act
+            // include the parent device as well to match
+            // IoT Hub API behavior
+            if (isModule)
+            {
+                Option<ServiceIdentity> device = await edgeHub.GetIdentityAsync(request.TargetDeviceId);
+                device.ForEach(i => identityList.Add(i));
             }
+
+            Events.SendingScopeResult(identityList);
+            result = MakeResultFromIdentities(identityList);
+            result.Status = HttpStatusCode.OK;
+
+            return result;
         }
 
-        internal static bool ValidateChainAndGetTargetDeviceId(string actorDeviceId, string authChain, out string targetDeviceId)
+        bool TryGetTargetDeviceId(string authChain, out string targetDeviceId)
         {
             targetDeviceId = string.Empty;
+
+            // The target device is the first ID in the provided authchain,
+            // which could be a module ID of the format "deviceId/moduleId".
             var actorAuthChainIds = authChain.Split(';', StringSplitOptions.RemoveEmptyEntries);
 
-            // Should have at least 1 element in the chain
-            if (actorAuthChainIds.Length < 1)
+            if (actorAuthChainIds.Length > 0)
             {
+                var ids = actorAuthChainIds[0].Split('/', StringSplitOptions.RemoveEmptyEntries);
+
+                if (ids.Length > 0)
+                {
+                    targetDeviceId = ids[0];
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        async Task<bool> AuthorizeActor(IEdgeHub edgeHub, string actorDeviceId, string actorModuleId, string targetId)
+        {
+            if (actorModuleId != Constants.EdgeHubModuleId)
+            {
+                // Only child EdgeHubs are allowed to act OnBehalfOf of devices/modules.
                 return false;
             }
 
-            // The actor device is the one that sent the request, so the last element
-            // of the auth-chain should always be the actor device's ID
-            if (!actorDeviceId.Equals(actorAuthChainIds.LastOrDefault()))
+            // Actor device is claiming to be our child, and that the target device is its child.
+            // So we should have an authchain already cached for the target device.
+            Option<string> targetAuthChain = await edgeHub.GetAuthChainForIdentity(targetId);
+
+            if (!targetAuthChain.HasValue)
             {
                 return false;
             }
 
-            // The target device ID is the first device ID in the chain, which could be a module ID
-            // of the format "deviceId/moduleId"
-            var ids = actorAuthChainIds[0].Split('/', StringSplitOptions.RemoveEmptyEntries).ToList();
-            targetDeviceId = ids[0];
+            // Validate the the target auth-chain
+            if (!AuthChainHelpers.ValidateAuthChain(actorDeviceId, targetId, targetAuthChain.Expect(() => new InvalidOperationException())))
```

Comment:
line 167 already validate targetAuthChain has value.  Use OrDefault or refactor to use ForEach.
```suggestion
            if (!AuthChainHelpers.ValidateAuthChain(actorDeviceId, targetId, targetAuthChain.OrDefault()))
```

Reply:
imo this is more defensive.  If the check on 167 gets changed/removed later, this will immediately throw an exception.  Whereas using OrDefault could potentially lead to a silent failure.

---
In reply to: [458526939](https://github.com/Azure/iotedge/pull/3248#discussion_r458526939) [](ancestors = 458526939)